In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
pd.__version__

'1.5.3'

In [3]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv


--2023-09-18 19:38:01--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘housing.csv’

housing.csv         100%[===================>]   1.36M  7.34MB/s    in 0.2s    

2023-09-18 19:38:02 (7.34 MB/s) - ‘housing.csv’ saved [1423529/1423529]



In [4]:
df = pd.read_csv('/content/housing.csv')

In [5]:
len(df.columns)

10

Question 2 - How many columns are in the dataset? - Answer - 10

In [9]:
missing_values = df.isna().sum()
missing_values[missing_values > 0]

total_bedrooms    207
dtype: int64

Question 3 - Which columns in the dataset have missing values? - total_bedrooms

In [11]:
len(df['ocean_proximity'].unique())

5

Question 4 - How many unique values does the ocean_proximity column have? - Answer - 5

In [13]:
df[df['ocean_proximity'] == 'NEAR BAY']['median_house_value'].mean()

259212.31179039303

Question 5 - What's the average value of the median_house_value for the houses located near the bay? - Answer - 259212

In [23]:
avg_bedrooms = df['total_bedrooms'].mean()
df['total_bedrooms'].fillna(avg_bedrooms, inplace = True)
new_avg_bedrooms = df['total_bedrooms'].mean()
print("Old average: ", avg_bedrooms)
print("New average: ", new_avg_bedrooms)

Old average:  537.8705525375617
New average:  537.8705525375617


Question 6 - Has it changed? - No

In [36]:
houses_on_island = df[df['ocean_proximity'] == 'ISLAND'][['housing_median_age', 'total_rooms', 'total_bedrooms']]
X = houses_on_island.values
XTX = np.matmul(X.T, X)
XTX_inv = np.linalg.inv(XTX)
y = [950, 1300, 800, 1000, 1300]
w = np.matmul(np.matmul(XTX_inv, X.T),y)
w

array([23.12330961, -1.48124183,  5.69922946])

Question 7 - What's the value of the last element of w? - Answer - 5.6992

This is the closed form solution of the linear regression.